File: sdks/java/src/main/java/com/blok/nanoservice/middleware/LoggingMiddleware.java
package com.blok.nanoservice.middleware;

import com.blok.nanoservice.logging.Logger;
import com.blok.nanoservice.node.NodeHandler;

import java.util.HashMap;
import java.util.Map;

/**
 * Middleware that logs node execution start and end with timing information.
 */
public class LoggingMiddleware implements Middleware {

    private final Logger logger;

    /**
     * Creates a LoggingMiddleware with the given logger.
     *
     * @param logger the logger to write to
     * @throws IllegalArgumentException if logger is null
     */
    public LoggingMiddleware(Logger logger) {
        if (logger == null) {
            throw new IllegalArgumentException("logger must not be null");
        }
        this.logger = logger;
    }

    @Override
    public NodeHandler wrap(NodeHandler next) {
        return (ctx, config) -> {
            long start = System.nanoTime();

            Map<String, Object> startFields = new HashMap<>();
            startFields.put("workflow", ctx.getWorkflowName());
            logger.info("node execution started", startFields);

            try {
                Object result = next.execute(ctx, config);

                long durationNanos = System.nanoTime() - start;
                double durationMs = durationNanos / 1_000_000.0;

                Map<String, Object> endFields = new HashMap<>();
                endFields.put("workflow", ctx.getWorkflowName());
                endFields.put("duration_ms", durationMs);
                logger.info("node execution completed", endFields);

                return result;

            } catch (Exception e) {
                long durationNanos = System.nanoTime() - start;
                double durationMs = durationNanos / 1_000_000.0;

                Map<String, Object> errorFields = new HashMap<>();
                errorFields.put("workflow", ctx.getWorkflowName());
                errorFields.put("duration_ms", durationMs);
                errorFields.put("error", e.getMessage());
                logger.error("node execution failed", errorFields);

                throw e;
            }
        };
    }
}

File: sdks/python3/blok/middleware/logging_middleware.py
from __future__ import annotations
import time
import logging
from typing import Any, Callable, Dict

logger = logging.getLogger("blok")


def logging_middleware(next_fn: Callable) -> Callable:
    """Middleware that logs the execution of each node with timing."""

    def wrapper(ctx, config: Dict[str, Any]) -> Any:
        start = time.monotonic()
        logger.info("node execution started [workflow=%s]", ctx.workflow_name)

        try:
            result = next_fn(ctx, config)
            duration_ms = (time.monotonic() - start) * 1000.0
            logger.info(
                "node execution completed [workflow=%s] [duration_ms=%.2f]",
                ctx.workflow_name,
                duration_ms,
            )
            return result
        except Exception as e:
            duration_ms = (time.monotonic() - start) * 1000.0
            logger.error(
                "node execution failed [workflow=%s] [duration_ms=%.2f] [error=%s]",
                ctx.workflow_name,
                duration_ms,
                str(e),
            )
            raise

    return wrapper
